sklearn/utils/sparsefuncs_fast.pyx

```diff
@@ -182,22 +201,87 @@ def _csc_mean_variance_axis0(np.ndarray[floating, ndim=1] X_data,
         startptr = X_indptr[i]
         endptr = X_indptr[i + 1]
         counts = endptr - startptr
+        n_samples_feat = n_samples
 
         for j in xrange(startptr, endptr):
-            means[i] += X_data[j]
-        means[i] /= n_samples
+            if isnan(X_data[j]) and ignore_nan:
+                n_samples_feat -= 1
+                counts -= 1
+        if n_samples_feat == 0: # Avoid division by Zero
+            continue
 
         for j in xrange(startptr, endptr):
-            diff = X_data[j] - means[i]
+            x_i = X_data[j]
+            if isnan(x_i) and ignore_nan:
+                continue
+            means[i] += x_i
+
+        means[i] /= n_samples_feat
+
+        for j in xrange(startptr, endptr):
+            x_i = X_data[j]
+            if isnan(x_i) and ignore_nan:
+                continue
+            diff = x_i - means[i]
             variances[i] += diff * diff
 
-        variances[i] += (n_samples - counts) * means[i] * means[i]
-        variances[i] /= n_samples
+        variances[i] += (n_samples_feat - counts) * means[i] * means[i]
+        variances[i] /= n_samples_feat
 
     return means, variances
 
 
-def incr_mean_variance_axis0(X, last_mean, last_var, unsigned long last_n):
+def n_samples_count_csc(np.ndarray[floating, ndim=1] X_data,
+                        shape,
+                        np.ndarray[int, ndim=1] X_indices,
+                        np.ndarray[int, ndim=1] X_indptr):
+    cdef unsigned int n_samples = shape[0]
+    cdef unsigned int n_features = shape[1]
+    cdef unsigned int startptr
+    cdef unsigned int endptr
+    cdef unsigned int i
+    cdef unsigned int j
+
+    cdef np.ndarray[unsigned int, ndim=1] n_samples_feat
+
+    n_samples_feat = np.ones(n_features, dtype=np.uint32) * n_samples
+
+    for i in xrange(n_features):
+        startptr = X_indptr[i]
+        endptr = X_indptr[i+1]
+
+        for j in xrange(startptr, endptr):
+            if isnan(X_data[j]):
+                n_samples_feat[i] -= 1
+
+    return n_samples_feat
+
+
+def n_samples_count_csr(np.ndarray[floating, ndim=1, mode="c"] X_data,
```

Comment:
I don't think we should need these helpers. I don't think we should need cython to count the number of NaNs in a column of a sparse matrix. Something like:
```
X = copy.copy(X)
X.data = np.isnan(X)
return X.sum(axis=0)
```
should suffice.

Reply:
@jnothman i was working on sparse matrices, but `np.isnan(X)` does not work for Sparse Matrices. I need to convert X into a dense numpy array using `X.toarray()` and then using `np.isnan(X)` or while make the count for NaN happen in the `for loops`.

I believe counting in `for loops` is better because these are sparse matrices and converting them to dense numpy array and then counting would probably make computations more expensive.